includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -49,22 +50,79 @@ public function __construct( $args = [] ) {
 	 * Register embed.
 	 */
 	public function register_embed() {
-		wp_embed_register_handler( 'amp-youtube', self::URL_PATTERN, [ $this, 'oembed' ], -1 );
-		add_shortcode( 'youtube', [ $this, 'shortcode' ] );
+		add_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10, 2 );
+		add_shortcode( 'youtube', [ $this, 'shortcode' ] ); // @todo Deprecated. See <https://github.com/ampproject/amp-wp/issues/3309>.
 		add_filter( 'wp_video_shortcode_override', [ $this, 'video_override' ], 10, 2 );
 	}
 
 	/**
 	 * Unregister embed.
 	 */
 	public function unregister_embed() {
-		wp_embed_unregister_handler( 'amp-youtube', -1 );
-		remove_shortcode( 'youtube' );
+		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
+		remove_shortcode( 'youtube' ); // @todo Deprecated. See <https://github.com/ampproject/amp-wp/issues/3309>.
 	}
 
+	/**
+	 * Filter oEmbed HTML for YouTube to convert to AMP.
+	 *
+	 * @param string $cache Cache for oEmbed.
+	 * @param string $url   Embed URL.
+	 * @return string Embed.
+	 */
+	public function filter_embed_oembed_html( $cache, $url ) {
+		$host = wp_parse_url( $url, PHP_URL_HOST );
+		if ( ! in_array( $host, [ 'youtu.be', 'youtube.com', 'www.youtube.com' ], true ) ) {
+			return $cache;
+		}
+
+		$id = $this->get_video_id_from_url( $url );
+		if ( ! $id ) {
+			return $cache;
+		}
+
+		$props = $this->parse_props( $cache, $url );
+
+		$props['video_id'] = $id;
+		return $this->render( $props, $url );
+	}
+
+	/**
+	 * Parse AMP component from iframe.
+	 *
+	 * @param string      $html HTML.
+	 * @param string|null $url  Embed URL, for fallback purposes.
+	 * @return array Props for rendering the component.
+	 */
+	private function parse_props( $html, $url ) {
+		$props = [];
+
+		if ( preg_match( '#<iframe[^>]*?title="(?P<title>[^"]+)"#s', $html, $matches ) ) {
+			$props['title']    = $matches['title'];
+			$props['fallback'] = sprintf(
+				'<a fallback href="%s">%s</a>',
```

Comment:
I keep arguing with myself whether this should be a `fallback` or a `placeholder`: https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/placeholders/

Actually, I think the answer is this _should_ be changed to rather be an `amp-img` as a `placeholder` as shown on the [`amp-youtube`](https://amp.dev/documentation/components/amp-youtube/) docs:

```html
  <amp-youtube
      id="myLiveChannel"
      data-live-channelid="UCB8Kb4pxYzsDsHxzBfnid4Q"
      width="358"
      height="204"
      layout="responsive">
    <amp-img
      src="https://i.ytimg.com/vi/Wm1fWz-7nLQ/hqdefault_live.jpg"
      placeholder
      layout="fill"
      />
  </amp-youtube>
```

The example there is incomplete however. In particular, the `amp-img` should have an `alt` attribute provided which is the `title` of the video.

Also, instead of actually generating an `amp-img` here, just output a regular `img` as it will get converted into an `amp-img` later by the image sanitizer.

Also, the `amp-img` placeholder would be suitable even when there is no `title` available, it's just that the `alt` would have to be blank.

Also, keeping the link to the original YouTube video is good. 

So what to do:

0. Unconditionally construct the link regardless of whether the `iframe` has a `title`.
1. Change this to be `placeholder` instead of `fallback`.
2. Parse construct the HQ video thumbnail from the video ID (e.g. `https://i.ytimg.com/vi/$video_id/hqdefault.jpg`).
3. Supply an `img` element as the child of the `a` link.
4. Supply the `title` of the `iframe` as the `img[alt]`, if available.

So in the end, it should look like this:

```html
<p>
    <amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270" title="Rebecca Black - Friday">
        <a placeholder href="https://www.youtube.com/watch?v=kfVsfOSbJY0">
            <img src="https://i.ytimg.com/vi/kfVsfOSbJY0/hqdefault.jpg" alt="Rebecca Black &#8211; Friday" data-amp-layout="fill">
        </a>
    </amp-youtube>
</p>
```

And in WordPress<5.2, the only difference would be no `alt` attribute.